packages/@uppy/core/src/Uppy.js

```diff
@@ -508,12 +509,17 @@ class Uppy {
       newFile = onBeforeFileAddedResult
     }
 
-    try {
-      const filesArray = Object.keys(files).map(i => files[i])
-      this.#restricter.validate(newFile, filesArray)
-    } catch (err) {
-      this.#informAndEmit(err, newFile)
-      throw err
+    this.#restricter.validate(newFile)
+
+    // Users are asked to re-select recovered files without data,
+    // and to keep the progress, meta and everthing else, we only replace said data
+    if (files[newFile.id] && files[newFile.id].isGhost) {
+      newFile = {
+        ...files[newFile.id],
+        data: fileDescriptor.data,
+        isGhost: false,
```

Comment:
Shouldn't it be true? If you meant false, can you please add a comment explaining why?
```suggestion
        isGhost: true,
```

Reply:
Oh I see now that this code was there before your change. Hum it's fine to ignore then, but it's a bit counter intuitive.

Reply:
Yea i dunno how that stuff works 😬 maybe @arturi knows

Reply:
I definitely agree we should add a comment when we find out

Reply:
Why can't we just replace the whole file? why do we need to only replace the data?

Reply:
Cause it can have meta from the user, for instance. Can't think of other reasons straight away.

Reply:
I removed that code and reselecting ghosts still work

Reply:
Ahh, maybe that's why this code exists huh? to preserve metadata. then i will add a comment about that